language-plutus-core/src/Language/PlutusCore/Normalize.hs

```diff
@@ -55,21 +66,33 @@ do so, because picked values cannot contain uninstantiated variables as only nor
 added to environments and normalization instantiates all variables presented in an environment.
 -}
 
+{- Note [Costs]
+Typechecking costs are relatively simple: it costs 1 gas to perform
+a substitution. Costs are reset every time we enter 'NormalizeTypeM'.
+
+In unlimited mode, gas is not tracked and we do not fail even on large numbers
+of reductions.
+-}
+
 -- See Note [Normalization].
 -- | Normalize a 'Type' in the 'NormalizeTypeM' monad.
 normalizeTypeM
     :: HasUnique (tyname ()) TypeUnique
-    => Type tyname () -> NormalizeTypeM tyname (NormalizedType tyname ())
-normalizeTypeM (TyForall ann name kind body) = TyForall ann name kind <<$>> normalizeTypeM body
-normalizeTypeM (TyFix ann name pat)          = TyFix ann name <<$>> normalizeTypeM pat
+    => Type tyname () -> NormalizeTypeM tyname a (NormalizedType tyname ())
+normalizeTypeM (TyForall ann name kind body) =
+    TyForall ann name kind <<$>> normalizeTypeM body
+normalizeTypeM (TyFix ann name pat)          =
+    TyFix ann name <<$>> normalizeTypeM pat
 normalizeTypeM (TyFun ann dom cod)           =
     TyFun ann <<$>> normalizeTypeM dom <<*>> normalizeTypeM cod
-normalizeTypeM (TyLam ann name kind body)    = TyLam ann name kind <<$>> normalizeTypeM body
+normalizeTypeM (TyLam ann name kind body)    =
+    TyLam ann name kind <<$>> normalizeTypeM body
 normalizeTypeM (TyApp ann fun arg)           = do
     vFun <- normalizeTypeM fun
     vArg <- normalizeTypeM arg
     case getNormalizedType vFun of
-        TyLam _ nArg _ body -> substituteNormalizeTypeM vArg nArg body
+        -- TODO: we should check nArg more thoroughly for non-normal types
```

Comment:
`nArg` is a type name, so I don't think it can have any non-normal types in it :)

Reply:
Still valid. I think you can just delete the comment.

Reply:
Will do.